src/system/group.cr

```diff
@@ -0,0 +1,195 @@
+require "c/grp"
+require "c/sys/limits"
+
+struct System::Group
+  module Limits
+    {% if flag?(:darwin) || flag?(:openbsd) || flag?(:freebsd) %}
+      GID_MAX = LibC::GID_MAX
+    {% elsif flag?(:linux) %}
+      GID_MAX = 0xffffffff_u32
+    {% else %}
+      GID_MAX = 0xffffffff_u32
+    {% end %}
+    NGROUPS_MAX = LibC::NGROUPS_MAX
+  end
+
+  # Converts group ID into a groupname.
+  #
+  # Returns: The groupname for the given group ID.
+  #
+  # ```
+  # System::Group.name(0) # => root
+  # ```
+  def self.name(gid : Int) : String
+    group_struct = LibC.getgrgid(int_to_gid(gid))
+    return String.new(group_struct.value.gr_name) if !group_struct.null?
+    raise NotFound.new("Group with gid '#{gid}', was not found.")
+  end
+
+  # Converts groupname into group ID.
+  #
+  # Returns: The group ID for the given groupname.
+  # Raises: Group::NotFound error if no group exists with the given groupname.
+  #
+  # ```
+  # System::Group.gid("root") # => 0
+  # ```
+  def self.gid(group : String) : Int
+    group_struct = LibC.getgrnam(group.check_no_null_byte)
+    return group_struct.value.gr_gid if !group_struct.null?
+    raise NotFound.new("Group with name '#{group.to_s}', was not found.")
+  end
+
+  # Checks to see if a group with the given group ID exists.
+  #
+  # Returns: `true` if the group exists.
+  #
+  # ```
+  # System::Group.exists?("root")              # => true
+  # System::Group.exists?("nonexistant_group") # => false
+  # ```
+  def self.exists?(gid : Int) : Bool
+    !LibC.getgrgid(int_to_gid(gid)).null?
+  end
+
+  # Checks to see if group with a given groupname exists.
+  #
+  # Returns: `true` if the group exists.
+  #
+  # ```
+  # System::Group.exists?(0)     # => true
+  # System::Group.exists?(32766) # => false
+  # ```
+  def self.exists?(group : String) : Bool
+    !LibC.getgrnam(group.check_no_null_byte).null?
+  end
+
+  # Returns the group specified by the group ID.
+  # ```
+  # System::Group.get(0)
+  # ```
+  def self.get(gid : Int) : Group
+    new(gid)
+  end
+
+  # Returns the group specified by the groupname.
+  # ```
+  # System::Group.get("root")
+  # ```
+  def self.get(groupname : String) : Group
+    new(groupname)
+  end
+
+  # Initializes a group by the given group ID.
+  def initialize(gid : Int)
+    group_struct = LibC.getgrgid(Group.int_to_gid(gid))
+    raise NotFound.new("Group with gid '#{gid}', was not found.") if group_struct.null?
+
+    group_struct = group_struct.value
+    @name = String.new(group_struct.gr_name)
+    @gid = group_struct.gr_gid
+    slice = group_struct.gr_mem.to_slice_null_terminated(LibC::NGROUPS_MAX)
+    @member_names = Array.new(slice.size) { |idx| String.new(slice[idx]) }
+  end
+
+  # Initializes a group by the given groupname.
+  def initialize(groupname : String)
+    group_struct = LibC.getgrnam(groupname.check_no_null_byte)
+    raise NotFound.new("Group with name '#{groupname}', was not found.") if group_struct.null?
```

Comment:
ditto

Reply:
Fixed to my standards. I will not use unless.